torchrl/collectors/collectors.py

```diff
@@ -322,10 +322,56 @@ def __init__(
         self._tensordict.set(
             "step_count", torch.zeros(*self.env.batch_size, 1, dtype=torch.int)
         )
-        self._tensordict_out = TensorDict(
-            {},
-            batch_size=[*self.env.batch_size, self.frames_per_batch],
-            device=self.passing_device,
+
+        # TODO: can we rely on policy having spec and ditch hasattr?
+        if (
+            hasattr(policy, "spec")
+            and policy.spec is not None
+            and all(v is not None for v in policy.spec.values())
+            and set(policy.spec.keys()) == set(policy.out_keys)
+        ):
+            # if policy spec is non-empty, all the values are not None and the keys
+            # match the out_keys we assume the user has given all relevant information
+            self._tensordict_out = env.reset()
+            self._tensordict_out.update(policy.spec.rand())
```

Comment:
I've sort of done this but without `unsqueeze`. I think what I have maybe isn't quite right, but ended up being necessary because scalar values specified via `UnboundedContinuousTensorSpec` already have dimension which meant I was getting a clash after unsqueezing.

However, now I have the spec for the hidden layers with shape `(1, hidden_size)` which doesn't seem quite right either.